source/Plugins/Language/Swift/SwiftCodeCompletion.cpp

```diff
@@ -192,8 +192,166 @@ doCodeCompletion(SourceFile &SF, StringRef EnteredCode,
+    PersistentExpressionState.RunOverHandLoadedModules(
+        [&](const ConstString ModuleName) -> bool {
+          // Skip these modules, to prevent "TestSwiftCompletions.py" from
+          // segfaulting.
+          // TODO: Investigate why this happens.
+          if (ModuleName == ConstString("SwiftOnoneSupport"))
+            return true;
+          if (ModuleName == ConstString("a"))
+            return true;
+
+          ModuleDecl *Module = SwiftCtx.GetModule(ModuleName, Error);
+          if (!Module)
+            return true;
+          if (ExistingImportSet.find(Module) != ExistingImportSet.end())
+            return true;
+          NewImports.push_back(SourceFile::ImportedModuleDesc(
+              std::make_pair(ModuleDecl::AccessPathTy(), Module),
+              SourceFile::ImportOptions()));
+        });
+    EnteredCodeFile->addImports(NewImports);
+  }
+
+  // We reset this file with the persistent decls every completion request.
+  SourceFile *PreviousDeclsFile =
+      GetSingleSourceFile(CompletionsModule, SourceFileKind::Library);
+  assert(PreviousDeclsFile);
+
+  // Reset the decls to the persistent decls.
+  {
+    std::vector<Decl *> PersistentDecls;
+    PersistentExpressionState.GetAllDecls(PersistentDecls);
+    PreviousDeclsFile->Decls.clear();
+    for (auto *PersistentDecl : PersistentDecls)
+      PreviousDeclsFile->Decls.push_back(PersistentDecl);
+    PreviousDeclsFile->clearLookupCache();
+  }
+
+  // `PreviousDeclsFile` might now contain decls that a re-defined in
+  // `EnteredCode`. We want to remove these so that the completion results only
+  // include results from the new definitions. To do this, we parse the
+  // `EnteredCode` to get a list of decls and then remove these decls from
+  // `PreviousDeclsFile`.
+  {
+    // Parse `EnteredCode` to populate `NewDecls`.
+    DenseMap<Identifier, SmallVector<ValueDecl *, 1>> NewDecls;
+    DiagnosticTransaction DelayedDiags(Ctx.Diags);
+    std::string AugmentedCode = EnteredCode.str();
+    AugmentedCode += '\0';
+    const unsigned BufferID =
+        Ctx.SourceMgr.addMemBufferCopy(AugmentedCode, "<REPL Input>");
+    const unsigned OriginalDeclCount = EnteredCodeFile->Decls.size();
+    PersistentParserState PersistentState(Ctx);
+    bool Done;
+    do {
+      parseIntoSourceFile(*EnteredCodeFile, BufferID, &Done, nullptr,
+                          &PersistentState, nullptr);
+    } while (!Done);
+    for (auto it = EnteredCodeFile->Decls.begin() + OriginalDeclCount;
+         it != EnteredCodeFile->Decls.end(); it++)
+      if (auto *NewValueDecl = dyn_cast<ValueDecl>(*it))
+        NewDecls.FindAndConstruct(NewValueDecl->getBaseName().getIdentifier())
+            .second.push_back(NewValueDecl);
+    EnteredCodeFile->Decls.resize(OriginalDeclCount);
+    DelayedDiags.abort();
+
+    // Subtract `NewDecls` from the decls in `PreviousDeclsFile`.
+    auto ContainedInNewDecls = [&](Decl *OldDecl) -> bool {
```

Comment:
Can't you use the swift_persistent_state->FindMatchingDecls / GetSwiftPersistentDecls to check if a decl is already there?

Reply:
We discussed this a bit in person.

In summary, this sounds good to do eventually. A current blocker is that `swift_persistent_state->FindMatchingDecls / GetSwiftPersistentDecls` requires the AST to be typechecked, but I need to do my deduping before typechecking. I think it might be possible to modify `swift_persistent_state->FindMatchingDecls / GetSwiftPersistentDecls` to work on non-typechecked ASTs, but I will do that later in a separate PR (and probably into master because it's a change that would also make sense in master).